ComparisonWithOrleans.md

```diff
@@ -0,0 +1,203 @@
+  
+  * In Akka the central abstraction for Location Transparency is the ActorRef: it is a self-contained serializable representation of an Actor location that enables other Actors to send messages to it. ActorRefs can also be sent as part of messages to introduce Actors to one another.
+  
+##### Automatic Scale-Out
+
+  * Orleans Grains can be configured as *stateless worker* which allows multiple activations to be active at any given time. The runtime manages the scaling factor in response to the workload.
+  
+  * In Akka cluster-aware routers offer the same functionality as the stateless worker mode of Orleans with the difference that here the provided abstraction exposes available resources to the program as the cluster grows but Akka itself does not contain the logic or ability to spin up additional nodes in response to increased load. This is left to other tools that manage server fleets.
+  
+#### Actor Interfaces
+
+  * The API provided by Orleans uses code generation in order to present Actor interactions exactly like normal method invocations—with the only difference that all methods are required to return their result asynchronously. Composition of results is performed using the async–await language features of C# and friends.
+  
+  * Akka distinguishes Actor interactions from normal method invocations by requiring the use of the ActorRef’s `!` or `tell` methods in order to send messages. These method return no result value, receiving a reply from the Actor entails providing an ActorRef where the response is sent to.
+
+In this area the differences between Akka and Orleans are very pronounced and mirror the different intent stated up-front: while Orleans aims for convenience and uses syntax that is familiar to OOP practitioners, Akka makes messaging very explicit and requires the user to define message classes. There is one instance of providing a similar model as async–await, namely the interaction between a PersistentActor and its Journal using the `persist` and `persistAsync` methods, but this API leads to a stronger coupling of the two components than would otherwise be the case. For a PersistentActor this is intended since it cannot function without the Journal, but for Grain interactions users now have to choose between blocking the grain in an `await` call (in a logical sense, but still!) or permitting other invocations to be processed concurrently “during” the `await` call—which leads to much of the same race conditions and problems that shared-state concurrency has. In Akka the interaction of multiple Actors is clearly demarcated by sending messages, and if and Actor ignores all incoming requests until a certain reply has been received then that is an explicit choice that the programmer makes.
```

Comment:
This difference in the fault recover approaches might be one of the most significant "real" differences between Orleans and Akka, but the details and implications might not initially be apparent to people not so familiar with .NET TPL. 

To slightly simplify things a little, I tend to think of the Akka / Erlang "let-it-crash" approach as __vertical__ error propagation where the exception is sent "up" the actor supervision tree until something handles it and does some appropriate recover action.

Orleans on the other hand uses "try-catch" approach [based on the .NET TPL / async await pattern] which is more of a __horizontal__ error propagation - the exception is sent back "along" the call chain until something handles it and takes some appropriate recover action (catch block).

These are both very valid and useful ways to implement error recovery in a complex distributed system, although I do not personally have enough experience with let-it-crash approach to be able to be able to articulate as a researcher when one approach might be considered objectively "better" than the other. 

This is one area where I think both projects have a lot to contribute to an interesting technical discussion, and lots of teaching / learning opportunities on both sides!

At the moment, the individual evaluation of "best" error handling approach seems to be largely driven by which programming language people are most familiar with. 
I do recognize & declare my own bias here by saying that I never want to go back to programming _without_ the async/await functionality in .NET :)
As a former Java programmer, the EA Orbit Project's port of async/await functionality to Java is certainly an interesting initiative I am watching closely. https://github.com/electronicarts/orbit/pull/29

However, it only takes a quick look at the amazing uptime stats for some of Ericsson's systems written in Erlang to know that the let-it-crash approach has lots of merit too!